Bug in Nvidia Tegra 3 driver

I'm trying to run LÖVE on an Ouya, which has an Nvidia Tegra 3. It supports OpenGL ES 2.0, which [doesn't support 3D textures](http://docs.gl/es3/glTexImage3D#versions). It also doesn't support the [ES 2.0 extension for 3D textures](https://www.khronos.org/registry/OpenGL/extensions/OES/OES_texture_3D.txt).

Currently, LÖVE checks to see if the extension is supported, and if it is, it aliases the functions for easier use later:

https://github.com/love2d/love/blob/0b06d0d1f9defaef0353523012acf918cbc01954/src/modules/graphics/opengl/OpenGL.cpp#L394-L403

On my Ouya, is doesn't matter whether these functions get alias or not, because both the regular versions and the OES versions are NULL.

As a result, I get a segfault during the boot process when LÖVE uses a 3D texture function for the first time:

```
<unknown> 0x0000000000000000
glad::glTexImage3D(unsigned int, int, int, int, int, int, int, unsigned int, unsigned int, void const*) gladfuncs.hpp:376
love::graphics::opengl::OpenGL::rawTexStorage(love::graphics::TextureType, int, love::PixelFormat, bool&, int, int, int) OpenGL.cpp:1171
love::graphics::opengl::OpenGL::createDefaultTexture() OpenGL.cpp:533
love::graphics::opengl::OpenGL::setupContext() OpenGL.cpp:261
love::graphics::opengl::Graphics::setMode(int, int, int, int, bool) Graphics.cpp:167
love::window::sdl::Window::setWindow(int, int, love::window::WindowSettings*) Window.cpp:527
love::window::w_setMode(lua_State*)::$_2::operator()() const wrap_Window.cpp:119
int love::luax_catchexcept<love::window::w_setMode(lua_State*)::$_2>(lua_State*, love::window::w_setMode(lua_State*)::$_2 const&) runtime.h:663
love::window::w_setMode(lua_State*) wrap_Window.cpp:118
lj_BC_FUNCC 0x0000000068df1ed0
love::luax_resume(lua_State*, int) runtime.cpp:971
runlove(int, char**, int&) love.cpp:218
::SDL_main(int, char **) love.cpp:259
Java_org_libsdl_app_SDLActivity_nativeRunMain SDL_android.c:761
dvmPlatformInvoke 0x0000000040749e34
dvmCallJNIMethod(unsigned int const*, JValue*, Method const*, Thread*) 0x0000000040778dce
dvmJitToInterpNoChain 0x00000000407532a8
```

If I comment out that line I get this segfault:

```
<unknown> 0x0000000000000000
glad::glTexSubImage3D(unsigned int, int, int, int, int, int, int, int, unsigned int, unsigned int, void const*) gladfuncs.hpp:377
love::graphics::opengl::OpenGL::createDefaultTexture() OpenGL.cpp:549
love::graphics::opengl::OpenGL::setupContext() OpenGL.cpp:261
love::graphics::opengl::Graphics::setMode(int, int, int, int, bool) Graphics.cpp:167
love::window::sdl::Window::setWindow(int, int, love::window::WindowSettings*) Window.cpp:527
love::window::w_setMode(lua_State*)::$_2::operator()() const wrap_Window.cpp:119
int love::luax_catchexcept<love::window::w_setMode(lua_State*)::$_2>(lua_State*, love::window::w_setMode(lua_State*)::$_2 const&) runtime.h:663
love::window::w_setMode(lua_State*) wrap_Window.cpp:118
lj_BC_FUNCC 0x0000000068df1db0
love::luax_resume(lua_State*, int) runtime.cpp:971
runlove(int, char**, int&) love.cpp:218
::SDL_main(int, char **) love.cpp:259
Java_org_libsdl_app_SDLActivity_nativeRunMain SDL_android.c:761
dvmPlatformInvoke 0x0000000040749e34
dvmCallJNIMethod(unsigned int const*, JValue*, Method const*, Thread*) 0x0000000040778dce
dvmJitToInterpNoChain 0x00000000407532a8
```

And finally, if I also comment out that line, I don't get a segfault, but LÖVE successfully catches the following fatal error:

```
Error: [string "boot.lua"]:535: Cannot compile vertex shader code:
    (104) : error C1115: unable to find compatible overloaded function "texture2DArray(sampler2DArray, vec3)"
    
    stack traceback:
    	[string "boot.lua"]:777: in function <[string "boot.lua"]:773>
    	[C]: in function 'setMode'
    	[string "boot.lua"]:535: in function <[string "boot.lua"]:380>
    	[C]: in function 'xpcall'
    	[string "boot.lua"]:787: in function <[string "boot.lua"]:780>
    	[C]: in function 'xpcall'
```

Awesome, thanks. It looks like `GLAD_OES_texture_3D == 0`, but `GLAD_EXT_texture_array == 1`. This lets textures of type TEXTURE_2D_ARRAY through here:

https://github.com/love2d/love/blob/0b06d0d1f9defaef0353523012acf918cbc01954/src/modules/graphics/opengl/OpenGL.cpp#L1217-L1218

I think the issue is that the [texture arrays extension](https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_texture_array.txt) doesn't come with support for the glTexImage3D, etc. functions. It can be used with those functions, but it doesn't define them.

When I do this, I get past the segfaults! :tada: 

```diff
diff --git a/src/modules/graphics/opengl/OpenGL.cpp b/src/modules/graphics/opengl/OpenGL.cpp
index f2fcbb01..12957f46 100644
--- a/src/modules/graphics/opengl/OpenGL.cpp
+++ b/src/modules/graphics/opengl/OpenGL.cpp
@@ -1210,7 +1210,7 @@ bool OpenGL::isTextureTypeSupported(TextureType type) const
        case TEXTURE_VOLUME:
                return GLAD_VERSION_1_1 || GLAD_ES_VERSION_3_0 || GLAD_OES_texture_3D;
        case TEXTURE_2D_ARRAY:
-               return GLAD_VERSION_3_0 || GLAD_ES_VERSION_3_0 || GLAD_EXT_texture_array;
+               return GLAD_VERSION_3_0 || GLAD_ES_VERSION_3_0 || GLAD_OES_texture_3D && GLAD_EXT_texture_array;
        case TEXTURE_CUBE:
                return GLAD_VERSION_1_3 || GLAD_ES_VERSION_2_0;
        default:
```

I still get the following crash though:

```
Error: [string "boot.lua"]:535: Cannot compile vertex shader code:
    (104) : error C1115: unable to find compatible overloaded function "texture2DArray(sampler2DArray, vec3)"
    
    stack traceback:
    	[string "boot.lua"]:777: in function <[string "boot.lua"]:773>
    	[C]: in function 'setMode'
    	[string "boot.lua"]:535: in function <[string "boot.lua"]:380>
    	[C]: in function 'xpcall'
    	[string "boot.lua"]:787: in function <[string "boot.lua"]:780>
    	[C]: in function 'xpcall'
```

--- a/src/modules/graphics/opengl/OpenGL.cpp
+++ b/src/modules/graphics/opengl/OpenGL.cpp
@@ -391,15 +391,23 @@ void OpenGL::initOpenGLFunctions()
 		}
 	}
 
-	if (GLAD_ES_VERSION_2_0 && GLAD_OES_texture_3D && !GLAD_ES_VERSION_3_0)
+	if (GLAD_ES_VERSION_2_0 && !GLAD_ES_VERSION_3_0)
 	{
-		// Function signatures don't match, we'll have to conditionally call it
-		//fp_glTexImage3D = fp_glTexImage3DOES;
-		fp_glTexSubImage3D = fp_glTexSubImage3DOES;
-		fp_glCopyTexSubImage3D = fp_glCopyTexSubImage3DOES;
-		fp_glCompressedTexImage3D = fp_glCompressedTexImage3DOES;
-		fp_glCompressedTexSubImage3D = fp_glCompressedTexSubImage3DOES;
-		fp_glFramebufferTexture3D = fp_glFramebufferTexture3DOES;
+		// The Nvidia Tegra 3 driver (used by Ouya) claims to support GL_EXT_texture_array but
+		// segfaults if you actually try to use it. OpenGL ES 2.0 devices should use OES_texture_3D.
+		// GL_EXT_texture_array is for desktops.
+		GLAD_EXT_texture_array = false;
+
+		if (GLAD_OES_texture_3D)
+		{
+			// Function signatures don't match, we'll have to conditionally call it
+			//fp_glTexImage3D = fp_glTexImage3DOES;
+			fp_glTexSubImage3D = fp_glTexSubImage3DOES;
+			fp_glCopyTexSubImage3D = fp_glCopyTexSubImage3DOES;
+			fp_glCompressedTexImage3D = fp_glCompressedTexImage3DOES;
+			fp_glCompressedTexSubImage3D = fp_glCompressedTexSubImage3DOES;
+			fp_glFramebufferTexture3D = fp_glFramebufferTexture3DOES;
+		}
 	}
 
 	if (!GLAD_VERSION_3_2 && !GLAD_ES_VERSION_3_2 && !GLAD_ARB_draw_elements_base_vertex)

--- a/src/modules/graphics/wrap_GraphicsShader.lua
+++ b/src/modules/graphics/wrap_GraphicsShader.lua
@@ -54,7 +54,8 @@ GLSL.SYNTAX = [[
 	#define DepthCubeImage samplerCubeShadow
 #endif
 #define extern uniform
-#ifdef GL_EXT_texture_array
+#if defined(GL_EXT_texture_array) && (!defined(GL_ES) || __VERSION__ > 100)
+#define texture_arrays_enabled
 #extension GL_EXT_texture_array : enable
 #endif
 #ifdef GL_OES_texture_3D
@@ -85,7 +86,7 @@ uniform LOVE_HIGHP_OR_MEDIUMP vec4 love_ScreenSize;
 
 GLSL.FUNCTIONS = [[
 #ifdef GL_ES
-	#if __VERSION__ >= 300 || defined(GL_EXT_texture_array)
+	#if __VERSION__ >= 300 || defined(texture_arrays_enabled)
 		precision lowp sampler2DArray;
 	#endif
 	#if __VERSION__ >= 300 || defined(GL_OES_texture_3D)
@@ -121,7 +122,7 @@ GLSL.FUNCTIONS = [[
 	#if __VERSION__ > 100 || defined(GL_OES_texture_3D)
 		vec4 Texel(sampler3D s, vec3 c) { return love_texture3D(s, c); }
 	#endif
-	#if __VERSION__ >= 130 || defined(GL_EXT_texture_array)
+	#if __VERSION__ >= 130 || defined(texture_arrays_enabled)
 		vec4 Texel(sampler2DArray s, vec3 c) { return love_texture2DArray(s, c); }
 	#endif
 	#ifdef PIXEL
@@ -130,7 +131,7 @@ GLSL.FUNCTIONS = [[
 		#if __VERSION__ > 100 || defined(GL_OES_texture_3D)
 			vec4 Texel(sampler3D s, vec3 c, float b) { return love_texture3D(s, c, b); }
 		#endif
-		#if __VERSION__ >= 130 || defined(GL_EXT_texture_array)
+		#if __VERSION__ >= 130 || defined(texture_arrays_enabled)
 			vec4 Texel(sampler2DArray s, vec3 c, float b) { return love_texture2DArray(s, c, b); }
 		#endif
 	#endif

